src/Hypsography/adaptspace.jl

```diff
@@ -0,0 +1,204 @@
+                # extrapolating from the bottom face level of the domain
+                2 * (cZ_column[v] - fZ_column[v])
+            elseif v == Nv + 1
+                # if this is the domain max face level compute the metric
+                # extrapolating from the top face level of the domain
+                2 * fZ_column[v] - cZ_column[v - 1]
+            else
+                cZ_column[v] - cZ_column[v - 1]
+            end
+            ∂x∂ξ = reconstruct_metric(local_geom.∂x∂ξ, f∇Z_column[v], Δz)
+            W = local_geom.WJ / local_geom.J
+            J = det(Geometry.components(∂x∂ξ))
+            face_column[v] = Geometry.LocalGeometry(coord, J, W * J, ∂x∂ξ)
+        end
+
+        # update center metrics
+        for v in 1:Nv
+            local_geom = center_column[v]
+            coord =
+                if typeof(space) <: Spaces.FaceExtrudedFiniteDifferenceSpace2D
+                    c1 = Geometry.components(local_geom.coordinates)[1]
+                    typeof(local_geom.coordinates)(c1, cZ_column[v])
+                else
+                    c1 = Geometry.components(local_geom.coordinates)[1]
+                    c2 = Geometry.components(local_geom.coordinates)[2]
+                    typeof(local_geom.coordinates)(c1, c2, cZ_column[v])
+                end
+            Δz = fZ_column[v + 1] - fZ_column[v]
+            ∂x∂ξ = reconstruct_metric(local_geom.∂x∂ξ, c∇Z_column[v], Δz)
+            W = local_geom.WJ / local_geom.J
+            J = det(Geometry.components(∂x∂ξ))
+            center_column[v] = Geometry.LocalGeometry(coord, J, W * J, ∂x∂ξ)
+        end
+    end
+
+    return Spaces.ExtrudedFiniteDifferenceSpace(
+        space.staggering,
+        space.horizontal_space,
+        TerrainWarpedIntervalTopology(space.vertical_topology),
+        space.global_geometry,
+        space.center_local_geometry,
+        space.face_local_geometry,
+        space.center_ghost_geometry,
+        space.face_ghost_geometry,
+    )
+end
+
+"""
+    adapt_space!(space, adaption, reference_distance)
+
+Inplace updates local_geometry coordinate and metric terms given a surface elevation
+`reference_distance` Field (ex. mean sea-level). Vertical level coordinates are adapted
+to the surface elevation warped space given a `TerrainAdaption` method.
+
+Returns the adapted `ExtrudedFiniteDifferenceSpace`.
+"""
+function adapt_space!(
+    space::Spaces.ExtrudedFiniteDifferenceSpace,
+    adaption::TerrainAdaption,
+    reference_distance::Fields.Field,
+)
+    # construct Z levels at all faces
+    if axes(reference_distance) !== space.horizontal_space
+        error(
+            "`reference_distance` space is not the same instance as the `ExtrudedFiniteDifferenceSpace` horizontal space",
+        )
+    end
+    coords =
+        Fields.coordinate_field(Spaces.FaceExtrudedFiniteDifferenceSpace(space))
+    zindex = length(coords |> propertynames)
+    vertical_ref = Geometry.component.(coords, zindex)
```

Comment:
```suggestion
    vertical_ref = if typeof(space) <: Spaces.FaceExtrudedFiniteDifferenceSpace2D 
        Geometry.components.(coords, 2)
    else
        Geometry.components.(coords, 3)
    end
```
would address your comment